src/wallet/wallet.cpp

```diff
@@ -4346,11 +4346,37 @@ util::Result<MigrationResult> MigrateLegacyToDescriptor(const std::string& walle
     // If the wallet is still loaded, unload it so that nothing else tries to use it while we're changing it
     bool was_loaded = false;
     if (auto wallet = GetWallet(context, wallet_name)) {
+        if (wallet->IsWalletFlagSet(WALLET_FLAG_DESCRIPTORS)) {
+            return util::Error{_("Error: This wallet is already a descriptor wallet")};
+        }
+
         if (!RemoveWallet(context, wallet, /*load_on_start=*/std::nullopt, warnings)) {
             return util::Error{_("Unable to unload the wallet before migrating")};
         }
         UnloadWallet(std::move(wallet));
         was_loaded = true;
+    } else {
+        // We need to load this wallet to check if there's something to migrate.
+        // But this cannot be reused later since we want to use BERKELEY_RO for the actual migration
+        WalletContext empty_context;
+        empty_context.args = context.args;
+        DatabaseOptions options;
+        options.require_existing = true;
+        DatabaseStatus status;
+        std::unique_ptr<WalletDatabase> database = MakeWalletDatabase(wallet_name, options, status, error);
+        if (!database) {
+            return util::Error{Untranslated("Wallet file verification failed.") + Untranslated(" ") + error};
+        }
+        wallet = CWallet::Create(empty_context, wallet_name, std::move(database), options.create_flags, error, warnings);
+        if (!wallet) {
+            return util::Error{Untranslated("Wallet loading failed.") + Untranslated(" ") + error};
+        }
+
+        if (wallet->IsWalletFlagSet(WALLET_FLAG_DESCRIPTORS)) {
+            return util::Error{_("Error: This wallet is already a descriptor wallet")};
+        }
+        // Unload
+        wallet.reset();
```

Comment:
Loading the entire wallet just to obtain the descriptors flag shouldn't be needed. Could just check the file type instead. We want a bdb file here. So, something like https://github.com/furszy/bitcoin-core/commit/7a03a039ae99e69504fe9df95b04f5e326c45a70 would work well. Yet, the code isn't the best but it includes a test that verifies the behavior.

Reply:
Good point, I think the original intention was to be able to handle the legacy-sqlite case, but that's moot since we require BDB_RO later. I've changed this to check for just look for BDB files, and also added your test.